File: framework/stt/whisper.py
import logging
import os
import queue
import tempfile

import numpy as np
import requests
import sounddevice as sd
import soundfile as sf

from ..config import ConfigManager
from .base import STTBackend


class WhisperSTT(STTBackend):
    """OpenAI Whisper API implementation of STT backend"""

    def __init__(self):
        self.config = ConfigManager()
        self.sample_rate = 16000

        self.service = self.config.get("stt.modules.whisper.service", "openai")
        service_config = self.config.get(
            f"stt.modules.whisper.{self.service}", {}
        )

        if self.service not in ["openai", "custom"]:
            raise ValueError(f"Unknown Whisper service: {self.service}")

        self.api_key = service_config.get("api_key")
        self.api_url = service_config.get("api_url")
        self.model = service_config.get("model", "whisper-1")
        self.headers = service_config.get("headers", {})

        if not self.api_key or not self.api_url:
            raise RuntimeError(
                f"Whisper {self.service} service not properly configured"
            )

    def listen(self) -> str:
        """Record audio from microphone and transcribe"""
        logger = logging.getLogger(__name__)
        logger.info("Listening... Press Enter to stop recording")

        # Setup audio queue and recording flag
        audio_queue = queue.Queue()
        recording = True

        def audio_callback(indata, frames, time, status):
            if status:
                logger.warning(f"Audio status: {status}")
            if recording:
                audio_queue.put(indata.copy())

        # Start recording in a separate thread
        stream = sd.InputStream(
            samplerate=self.sample_rate,
            channels=1,
            callback=audio_callback,
            blocksize=int(self.sample_rate * 0.5),  # 0.5 second blocks
        )

        chunks = []
        with stream:
            # Wait for Enter key
            input()
            recording = False

            # Get remaining data from queue
            while not audio_queue.empty():
                chunks.append(audio_queue.get())

        logger.info("Processing speech...")

        if not chunks:
            logger.warning("No audio recorded")
            return ""

        # Convert to numpy array
        audio_data = np.concatenate(chunks)

        # Save to temporary file
        with tempfile.NamedTemporaryFile(suffix=".wav", delete=False) as f:
            sf.write(f.name, audio_data, self.sample_rate)
            return self.transcribe_file(f.name)

    def transcribe_file(self, audio_file: str) -> str:
        """Transcribe an audio file using configured Whisper service"""
        try:
            headers = self.headers.copy()
            if self.service == "openai":
                headers["Authorization"] = f"Bearer {self.api_key}"
            elif self.service == "custom":
                headers["Authorization"] = self.api_key

            with open(audio_file, "rb") as f:
                files = {
                    "file": (os.path.basename(audio_file), f, "audio/wav"),
                    "model": (None, self.model),
                    "response_format": (None, "json"),
                    "language": (None, "auto"),
                    "task": (None, "transcribe"),
                }

                response = requests.post(
                    self.api_url, headers=headers, files=files
                )

                response.raise_for_status()
                result = response.json()

                return result.get("text", "").strip()

        except Exception as e:
            logger = logging.getLogger(__name__)
            logger.error(f"Transcription failed: {str(e)}")
            raise RuntimeError(
                f"Whisper {self.service} transcription failed: {str(e)}"
            )

File: polaris/services/stt/whisper.js
// Ainara AI Companion Framework Project
// Copyright (C) 2025 Rubén Gómez - khromalabs.org
//
// This file is dual-licensed under:
// 1. GNU Lesser General Public License v3.0 (LGPL-3.0)
//    (See the included LICENSE_LGPL3.txt file or look into
//    <https://www.gnu.org/licenses/lgpl-3.0.html> for details)
// 2. Commercial license
//    (Contact: rgomez@khromalabs.org for licensing options)
//
// You may use, distribute and modify this code under the terms of either license.
// This notice must be preserved in all copies or substantial portions of the code.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
// Lesser General Public License for more details.

const STTBackend = require('./base');
const ConfigManager = require('../../framework/config');
const ConfigHelper = require('../../framework/ConfigHelper');
const { ipcRenderer } = require('electron');
const { promisify } = require('util');
const fs = require('fs');
const os = require('os');
const path = require('path');

const ffmpeg = require('fluent-ffmpeg');
const ffmpegPath = require('ffmpeg-static');
ffmpeg.setFfmpegPath(ConfigHelper.resolveBinaryPath(ffmpegPath));



class WhisperSTT extends STTBackend {
    /**
     * OpenAI Whisper API implementation of STT backend
     */
    constructor() {
        super();

        const config = new ConfigManager();
        this.service = config.get('stt.modules.whisper.service', 'openai');
        const serviceConfig = config.get(`stt.modules.whisper.${this.service}`, {});

        if (!['openai', 'pybridge', 'custom'].includes(this.service)) {
            throw new Error(`Unknown Whisper service: ${this.service}`);
        }

        this.apiKey = serviceConfig.apiKey;
        this.apiUrl = serviceConfig.apiUrl;
        this.model = serviceConfig.model || 'whisper-1';
        this.headers = serviceConfig.headers || {};

        if (!this.apiKey || !this.apiUrl) {
            throw new Error(`Whisper ${this.service} service not properly configured`);
        }
    }

    async initialize() {
        try {
            console.log("WhisperSTT initialize starting");
            console.log("apiUrl is:", this.apiUrl);
            console.log("ffmpegPath is:", ConfigHelper.resolveBinaryPath(ffmpegPath));
            // const url = new URL(this.apiUrl);
            // // Check if server is available by requesting the root URL
            // const baseUrl = `${url.protocol}//${url.host}`;
            const baseUrl = this.apiUrl;
            console.log("Checking Whisper server at: " + baseUrl);
            const response = await fetch(baseUrl);
            if (!response.ok) {
                throw new Error(`Whisper service not available at ${baseUrl}`);
            }
            console.log('Whisper service connection established');
        } catch (error) {
            const msg = 'Speech-to-Text service is not available.\n\n' +
                'Please ensure the Whisper server is running and try again.\n\n' +
                'Error details: ' + error.message
            ipcRenderer.send('critical-error', msg);
            throw new Error('Whisper STT service is not available. Please ensure the Whisper server is running at ' + this.apiUrl);
        }
    }

    // In WhisperSTT class:
    async webmToWav(webmBlob) {
        try {
            console.log('Converting WebM to WAV...');

            // Create temp files
            const tempDir = os.tmpdir();
            const inputPath = path.join(tempDir, `input-${Date.now()}.webm`);
            const outputPath = path.join(tempDir, `output-${Date.now()}.wav`);

            // Write WebM blob to temp file
            const buffer = Buffer.from(await webmBlob.arrayBuffer());
            await promisify(fs.writeFile)(inputPath, buffer);

            // Convert using FFmpeg
            await new Promise((resolve, reject) => {
                ffmpeg(inputPath)
                    .toFormat('wav')
                    .audioFrequency(16000)  // Force 16 kHz sample rate
                    .audioChannels(1)       // Mono audio
                    .audioBitrate('16k')    // 16-bit depth
                    .on('error', reject)
                    .on('end', resolve)
                    .save(outputPath);
            });

            // Read result
            const wavData = await promisify(fs.readFile)(outputPath);
            const wavBlob = new Blob([wavData], { type: 'audio/wav' });

            // Cleanup temp files
            await promisify(fs.unlink)(inputPath);
            await promisify(fs.unlink)(outputPath);

            console.log(`WAV conversion complete, size: ${wavBlob.size} bytes`);
            return wavBlob;

        } catch (error) {
            console.error('Error converting WebM to WAV:', error);
            throw error;
        }
    }


    /**
     * Transcribes an audio blob.
     * Automatically converts WebM to WAV if needed.
     * @param {Blob} audioBlob
     * @returns {Promise<string>} Transcription text
     */
    async transcribe(audioBlob) {
        try {
            console.log(`WhisperSTT: Transcribing blob of type ${audioBlob.type}, size: ${audioBlob.size}`);

            let blobToProcess = audioBlob;

            // Convert if it's WebM (common from MediaRecorder)
            if (audioBlob.type.includes('webm') || audioBlob.type.includes('x-matroska')) {
                console.log('WhisperSTT: Detected WebM/Matroska, converting to WAV...');
                blobToProcess = await this.webmToWav(audioBlob);
            }

            return await this.transcribeFile(blobToProcess);
        } catch (error) {
            console.error('WhisperSTT: Transcription error:', error);
            throw error;
        }
    }

    async transcribeFile(audioBlob) {
        try {
            const headers = { ...this.headers };

            if (this.service === 'openai') {
                headers.Authorization = `Bearer ${this.apiKey}`;
            } else if (this.service === 'custom') {
                headers.Authorization = this.apiKey;
            }

            const formData = new FormData();
            formData.append('file', new File([audioBlob], 'recording.wav', { type: 'audio/wav' }));
            formData.append('model', this.model);
            formData.append('response_format', 'json');
            formData.append('language', 'auto');
            formData.append('task', 'transcribe');

            console.log("sending audio to: " + this.apiUrl)

            const response = await fetch(`${this.apiUrl}`, {
                method: 'POST',
                headers,
                body: formData
            });

            if (!response.ok) {
                throw new Error(`Server error: ${response.status}`);
            }

            const result = await response.json();
            console.log('Transcription:', result);
            
            // [MODIFIED] Return object with text and confidence
            return {
                text: result.text?.trim() || '',
                confidence: result.confidence, // May be undefined if using direct OpenAI
                language: result.language
            };

        } catch (error) {
            console.error(`Whisper ${this.service} transcription failed:`, error);
            throw error;
        }
    }
}

module.exports = WhisperSTT;
